src/devices/Charlieplex/Charlieplex.cs

```diff
@@ -0,0 +1,175 @@
+            }
+
+            var charlieCount = (int)Math.Pow(pins.Length, 2) - pins.Length;
+            if (nodeCount > charlieCount)
+            {
+                throw new ArgumentException($"{nameof(Charlieplex)}: maximum count is {charlieCount} based on {pins.Length} pins. {nodeCount} was specified as the count.");
+            }
+
+            if (nodeCount == 0)
+            {
+                nodeCount = charlieCount;
+            }
+
+            if (gpioController == null)
+            {
+                gpioController = new GpioController();
+            }
+
+            foreach (var pin in pins)
+            {
+                gpioController.OpenPin(pin, PinMode.Input);
+            }
+
+            _controller = gpioController;
+            _shouldDispose = shouldDispose;
+            _pins = pins;
+            _nodeCount = nodeCount;
+            _charlieNodes = GetCharlieLoads(pins, nodeCount);
+        }
+
+        /// <summary>
+        /// The number of load devices (like LEDs) that can be addressed.
+        /// </summary>
+        public int LoadCount => _nodeCount;
+
+        /// <summary>
+        /// Write a PinValue to a load.
+        /// Address scheme is 0-based. Given 8 loads, addresses would be 0-7.
+        /// </summary>
+        public void Write(int node, PinValue value, int delay = 10)
+        {
+            _charlieNodes[node].Value = value;
+
+            for (int i = 0; i < delay; i++)
+            {
+                WriteSegment();
+            }
+        }
+
+        private void WriteSegment()
+        {
+            // int delay = 1;
+            // configure the circuit back to a neutral configuration
+            foreach (var p in _pins)
+            {
+                _controller.SetPinMode(p, PinMode.Input);
+            }
+
+            for (int i = 0; i < _charlieNodes.Length; i++)
+            {
+                var node = _charlieNodes[i];
+                if (node.Value == PinValue.Low)
+                {
+                    continue;
+                }
+
+                // set ground on Cathode leg
+                _controller.SetPinMode(node.Cathode, PinMode.Output);
+                _controller.Write(node.Cathode, 0);
+
+                // set power on Anode leg
+                _controller.SetPinMode(node.Anode, PinMode.Output);
+                _controller.Write(node.Anode, 1);
+                // Thread.Sleep(delay);
+                _controller.SetPinMode(node.Anode, PinMode.Input);
+                _controller.SetPinMode(node.Cathode, PinMode.Input);
+            }
+
+        }
+
+        /// <summary>
+        /// Provides the set of Charlie loads given the set of pins and the count provided.
+        /// If count = 0, then the Charlieplex maximum is used for the pins provided (n^2-n).
+        /// </summary>
+        public static CharlieplexNode[] GetCharlieLoads(int[] pins, int loadCount = 0)
```

Comment:
Does this really need to be public? Inputs should be validated in the Ctor. Maybe instead offer a property to retrieve the _carlieNodes member?

Reply:
This makes it so that someone can call this API on any machine to get the correct wiring layout. That's useful. It's also what the tests uses for the wiring layout, again, runnable on any machine. The issue is that the constructor needs a GpioController, which only works on a Pi and similar devices.

Reply:
As I read docs / blog posts for other solutions, I found their description of how to wire the circuit confusing at times, or not super well connected to the code. This API is intended to solve that. It also covers the case where you want to enable fewer nodes than the max for the number of pins.

Reply:
Sounds reasonable then. Just maybe prefer List<CharliplexNode> as return type instead of an array?